lib/geocoder_api_client.py

```diff
@@ -0,0 +1,117 @@
+
+    def __init__(self):
+        self.logger = create_log('geocoder_api_client')
+
+        retry_policy = Retry(total=2, backoff_factor=4, status_forcelist=[
+                             500, 502, 503, 504], allowed_methods=frozenset(['GET', 'POST']))
+        self.session = requests.Session()
+        self.session.mount('https://', HTTPAdapter(max_retries=retry_policy))
+
+    def get_geoids(self, address_df):
+        """
+        Geocodes the addresses in address_df by sending two requests to the 
+        API: one with all of the input addresses and another with the addresses
+        that failed to be geocoded in the first request. This is recommended by
+        the API because it sometimes erroneously fails to match an address when
+        it's used in batch mode. For more info, see
+        https://www2.census.gov/geo/pdfs/maps-data/data/Census_Geocoder_FAQ.pdf
+
+        Returns a dataframe containing the geoids (or NaN) indexed to match
+        address_df.
+        """
+
+        self.logger.info('Sending batch of addresses to geocoder API')
+        geoids = self._get_geoids_with_single_request(address_df)
+        retry_indices = geoids[geoids.isnull()].index
+        retry_address_df = address_df.loc[retry_indices]
+        geoids.update(self._get_geoids_with_single_request(retry_address_df))
+        return geoids
+
+    def _get_geoids_with_single_request(self, address_df):
+        """
+        Geocodes the addresses in address_df using a single API request.
+
+        Returns a dataframe containing the geoids (or NaN) indexed to match
+        address_df.
+        """
+
+        with BytesIO() as address_stream:
+            address_df.to_csv(address_stream, header=False,
+                              columns=['address', 'city', 'region', 'postal_code'])
+            raw_response = self._send_request(address_stream)
+        response_df = pd.read_csv(BytesIO(raw_response), header=None, dtype=str,
+                                  index_col=0, engine='python', names=[
+                                  'index', 'input_address', 'match', 'match_type',
+                                  'matched_address', 'coordinates', 'tigerline_id',
+                                  'tigerline_side', 'state_id', 'county_id',
+                                  'tract_id', 'block_id'])
+        geoids = (response_df['state_id'] + response_df['county_id'] +
+                  response_df['tract_id']).rename('geoid')
+
+        if len(geoids) != len(address_df):
+            self.logger.error('Size of input to geocoder ({input_size}) does not match size of output ({output_size})'
+                              .format(input_size=len(address_df), output_size=len(geoids)))
+            raise GeocoderApiClientError('Size of input to geocoder ({input_size}) does not match size of output ({output_size})'
+                                         .format(input_size=len(address_df), output_size=len(geoids)))
```

Comment:
What's the implication of this error? That the geocoder was unable to process certain addresses?

Reply:
Is it necessary because otherwise we can't align the input and output dataframes?

Reply:
Removed. Originally I meant for it to check if the geocoder missed some addresses, but it turns out that never happens. The API either geocodes all the addresses in the batch or none if there is an error.